### ISSUE TITLE
Retry on `RST_STREAM with code 0` error

### ISSUE THREAD

**Issue by fundthmcalculus** [Dentity-Development/dentity-ecosystems-sdk#388]
This can be the result of intermittent network issues, memory pressure, etc.
https://stackoverflow.com/questions/69590889/js-sdk-v2-0-16-error-13-internal-received-rst-stream-with-code-0
https://github.com/grpc/grpc-node/issues/1532
we should just retry (probably have some limit how many).

Other than just wrapping every call in a while loop, thoughts on how to implement this?

**Comment by fundthmcalculus** [comment]
Discussed with Lucas, he will start on this 2022-01-25.

**Comment by tmarkovski** [comment]
Is this happening intermittently or consistently?

**Comment by fundthmcalculus** [comment]
> Is this happening intermittently or consistently?

Pretty consistently on my local machine, never on the CI pipeline, so it's a lower priority.

**Comment by tmarkovski** [comment]
Aren't we concerned its failing consistently locally? Is it specific to Java only?

**Comment by fundthmcalculus** [comment]
> Aren't we concerned its failing consistently locally? Is it specific to Java only?

It's Javascript only. And it's consistent, but random. And nobody else I know has the issue, so it's a lower priority.

**Comment by lucasamonrc** [comment]
I think our new client, Dentity, is blocked in their development because they can't get past this.

**Comment by fundthmcalculus** [comment]
I'll prioritize this then.

**Comment by tmarkovski** [comment]
How to reproduce this? If I run the [node example](https://github.com/trinsic-id/sdk-examples/tree/main/node/basic) here, can I get it reliably? Can we collect more details, which endpoints produce this? Under what conditions, etc.

**Comment by fundthmcalculus** [comment]
> 

Let me see. I see it on the node unit tests. I can consistently reproduce it locally. I'll work on this, just give me a couple days. Sorry.

**Comment by lucasamonrc** [comment]
I might have something actually. I happens every time I do this:
```ts
const accountService = new AccountService({ server });
const details = new AccountDetails();

details.setName("Lucas");
details.setEmail("lucasamonrc@gmail.com");

let lucas = (await accountService.signIn(details)).getProfile();
accountService.updateActiveProfile(lucas);

const info = await accountService.info(); // this line seems to be the problem.

console.log(info);
```

I have no problems when I remove this line:
```ts
const info = await accountService.info(); // this line seems to be the problem.
```

**Comment by lucasamonrc** [comment]
Furthermore, it works normally if I don't set any account details.

```ts
const accountService = new AccountService({ server });

let lucas = (await accountService.signIn(details)).getProfile();
accountService.updateActiveProfile(lucas);

const info = await accountService.info(); // doesn't seem to work with set account details
console.log(info);
```

**Comment by tmarkovski** [comment]
@lucasamonrc I think the problem is with using a profile that's been protected. When you pass an email in the sign in details, you get back a profile that has protection on. This can be witnessed if you inspect this response. When this happens, the profile must be `unprotected` first, using the code sent to the email. There should be `unprotect` function in the account service.
Once you unprotect the profile, you then use it in `accountService.updateActiveProfile()`, after which `info` and other service calls should work fine.

We need a better error message from the server to indicate this if possible, but also the SDK should throw an error when using unprotected token.

**Comment by lucasamonrc** [comment]
Well, I tried that as well. This my actual code:
```ts
const accountService = new AccountService({ server });
const details = new AccountDetails();

details.setName("Lucas");
details.setEmail("lucasamonrc@gmail.com");

let lucas = (await accountService.signIn(details)).getProfile();

const code = await askSecurityCode(); // this my own function I wrote so I could enter the code I got in my email.

lucas = await accountService.unprotect(lucas, code);
accountService.updateActiveProfile(lucas);

const info = await accountService.info(); // still have the same problem. once removed it worked.

console.log(info);
```

**Comment by allibell** [comment]
fyi I'm also seeing this when I persist the profile in a `.bin` using the Node SDK.  versus when I delete that `.bin`/create a fresh `AccountProfile` in memory, I don't see the error.  this seems to suggest you can use the profile (without unprotecting first) when you have first created it?  (not sure though, #noob)

**Comment by tmarkovski** [comment]
Some server exceptions I'm seeing related to this

```
info: Trinsic.Server.Security.Oberon.OberonAuthenticationHandler[7]
      Oberon was not authenticated. Failure message: Status(StatusCode="Unavailable", Detail="Error starting gRPC call. HttpRequestException: An error occurred while sending the request. IOException: The request was aborted. IOException: The response ended prematurely while waiting for the next frame from the server.", DebugException="System.Net.Http.HttpRequestException: An error occurred while sending the request.
       ---> System.IO.IOException: The request was aborted.
       ---> System.IO.IOException: The response ended prematurely while waiting for the next frame from the server.
         at System.Net.Http.Http2Connection.<ReadFrameAsync>g__ThrowMissingFrame|57_1()
         at System.Net.Http.Http2Connection.ReadFrameAsync(Boolean initialFrame)
         at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
         --- End of inner exception stack trace ---
         at System.Net.Http.Http2Connection.ThrowRequestAborted(Exception innerException)
         at System.Net.Http.Http2Connection.Http2Stream.CheckResponseBodyState()
         at System.Net.Http.Http2Connection.Http2Stream.TryEnsureHeaders()
         at System.Net.Http.Http2Connection.Http2Stream.ReadResponseHeadersAsync(CancellationToken cancellationToken)
         at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
         --- End of inner exception stack trace ---
         at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
         at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
         at System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
         at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
         at Grpc.Net.Client.Internal.GrpcCall`2.RunCall(HttpRequestMessage request, Nullable`1 timeout)")
```


```
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMG0BJJRA4D1", Request id "0HMG0BJJRA4D1:00000083": An unhandled exception was thrown by the application.
      System.InvalidOperationException: Invalid non-ASCII or control character in header: 0x000A
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.ThrowInvalidHeaderCharacter(Char ch)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.ValidateHeaderValueCharacters(String headerName, StringValues headerValues, Func`2 encodingSelector)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpResponseTrailers.SetValueFast(String key, StringValues value)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.Microsoft.AspNetCore.Http.IHeaderDictionary.set_Item(String key, StringValues value)
         at Microsoft.AspNetCore.Http.ParsingHelpers.SetHeaderUnmodified(IHeaderDictionary headers, String key, Nullable`1 values)
         at Microsoft.AspNetCore.Http.ParsingHelpers.AppendHeaderUnmodified(IHeaderDictionary headers, String key, StringValues values)
         at Microsoft.AspNetCore.Http.ResponseTrailerExtensions.AppendTrailer(HttpResponse response, String trailerName, StringValues trailerValues)
         at Trinsic.Server.Security.RpcExceptionFormatter.InvokeAsync(HttpContext context, RequestDelegate next) in /app/server-api/Server/Security/RpcExceptionFormatter.cs:line 26
         at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass6_1.<<UseMiddlewareInterface>b__1>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
```

**Comment by tmarkovski** [comment]
Tagging @sethjback for visibility.

We used this issue to track the RST_STREAM reset that happens on Node. Per our convo the other day, this issue is not network related and root cause is a crash in auth service because of proto default value handling. Seth mentioned there's a fix we can do to correct this.

**Comment by sethjback** [comment]
There are two outstanding tickets on the backend that will address this - one is a workaround for the underlying Oberon bug, and the other will address returning an error rather than simply closing the stream.

**Comment by lucasamonrc** [comment]
What the status update on this issue?

**Comment by sethjback** [comment]
We have a fix deployed in dev and staging that will be pushed to prod as part of the 1.4 sdk update.

**Comment by fundthmcalculus** [comment]
Fixed with the 1.4 release over the weekend.
